File: 2022/02/05/c/pathSum.c
// Program: https://leetcode.com/problems/path-sum/
// Author: Andrew Jarombek
// Date: 2/5/2022

#include <stddef.h>
#include <stdbool.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

bool dfs(struct TreeNode* root, int targetSum, int currentSum) {
    int sum = root->val + currentSum;
    if (root->left == NULL && root->right == NULL) {
        return sum == targetSum;
    }

    bool left = false;
    bool right = false;

    if (root->left != NULL) {
        left = dfs(root->left, targetSum, sum);
    }

    if (root->right != NULL) {
        right = dfs(root->right, targetSum, sum);
    }

    return left || right;
}

bool hasPathSum(struct TreeNode* root, int targetSum){
    if (root == NULL) {
        return false;
    }

    return dfs(root, targetSum, 0);
}

File: 2022/03/13/pathSum.swift
// https://leetcode.com/problems/path-sum/
// Author: Andrew Jarombek
// Date: 3/13/2022

class Solution {
    func hasPathSum(_ root: TreeNode?, _ targetSum: Int) -> Bool {
        if root == nil {
            return false
        }

        return dfs(root!, targetSum, 0)
    }

    func dfs(_ root: TreeNode, _ targetSum: Int, _ actualSum: Int) -> Bool {
        let newSum = actualSum + root.val

        if root.left == nil && root.right == nil {
            return newSum == targetSum
        }

        var left = false
        var right = false

        if let leftRoot = root.left {
            left = dfs(leftRoot, targetSum, newSum)
        }

        if let rightRoot = root.right {
            right = dfs(rightRoot, targetSum, newSum)
        }

        return left || right
    }
}